### ISSUE TITLE
Low text recognition accuracy for long words

### ISSUE THREAD

**Issue by Razlaw** [mindee/doctr#1936]
### Bug description

First of all, thanks a lot for creating and maintaining docTR :)
It's a really nice open source package and it looks like you are putting a lot of work into it.

I noticed some inconsistencies when using the RecognitionPredictor on long words.
Here is an example crop from a receipt:

![Image](https://github.com/user-attachments/assets/315bd653-c6b2-499b-8b55-a9c3e6d4b81b)

And here the same crop right before the text recognition:

![Image](https://github.com/user-attachments/assets/85128e2d-7d4a-4cca-847b-bb9b1f9070b8)

I saw in an old issue #521 that the proposed workaround was to increase the target aspect ratio for the splitting of the crops.
This didn't help though.
Actually the crop isn't split at all.

### Code snippet to reproduce the bug

--

### Error traceback

--

### Environment

DocTR version: 0.11.1a0
TensorFlow version: N/A
PyTorch version: 2.7.0+cu126 (torchvision 0.22.0+cu126)
OpenCV version: 4.11.0
OS: Ubuntu 22.04.5 LTS
Python version: 3.10.12
Is CUDA available (TensorFlow): N/A
Is CUDA available (PyTorch): No
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect

### Deep Learning backend

Both not available, running on CPU only.

**Comment by Razlaw** [comment]
Btw here is the result:

![Image](https://github.com/user-attachments/assets/2a3bfccb-52f7-4dbe-a074-c48572f4b883)

I'm currently looking deeper into the code and the example crop should trigger the splitting..

**Comment by Razlaw** [comment]
So, apparently this line should calculate the number of subcrops the crop should be split into
https://github.com/mindee/doctr/blob/db6d0dbe9aef671695b0f6d725196bec3c7d9a7d/doctr/models/recognition/predictor/_utils.py#L41
Putting in the values for the example we have an aspect_ratio of roughly 11.37 = 1126 / 99 and a default target_ratio of 6, I would expect to get 2 subcrops, but the result is 1 = 11.37 // 6.
But even when I adapted this, results didn't improve really.
Should I continue investigating or am I completely on the wrong track? :)

**Comment by felixdittrich92** [comment]
Hi @Razlaw 👋,

You're right - in Python, `//` performs floor division (rounds down), so the original behavior is correct. However, it might actually make more sense to do the opposite using:

```python
num_subcrops = math.ceil(aspect_ratio / target_ratio)
```

This ensures we don’t underestimate the number of sub-crops, which can help with better coverage.

In your specific case, I'd also recommend setting `assume_straight_pages=False` to improve recognition accuracy.

Here’s a minimal example for inference:

```python
from doctr.models import recognition_predictor
from doctr.io import DocumentFile

doc = DocumentFile.from_images(["/home/felix/Desktop/WORK/doctr/test.png"])
predictor = recognition_predictor(pretrained=True)

predictor.critical_ar = 8  # Critical aspect ratio
predictor.dil_factor = 1.4  # Dilation factor to overlap the crops
predictor.target_ar = 3  # Target aspect ratio # NOTE: changed from 6 to 3

res = predictor(doc)
print(res)  # Example output: 'XUrA+b3iHQP+F3qxu0zzlZZDRA/6Z8+6nbf'
```

That said, these are hyperparameters and there’s currently no direct or universal way to auto-calculate them per batch. What you could explore is dynamically computing these values per crop - for instance, by analyzing each image’s aspect ratio and choosing `target_ratio`, `max_ratio`, and `dilation` accordingly.

Here’s a note in the `split_crops` function where this could be done:

```python
def split_crops(
  crops: list[np.ndarray],
  max_ratio: float,
  target_ratio: int,
  dilation: float,
  channels_last: bool = True,
) -> tuple[list[np.ndarray], list[int | tuple[int, int]], bool]:
  """Chunk crops horizontally to match a given aspect ratio

  Args:
      crops: list of numpy array of shape (H, W, 3) if channels_last or (3, H, W) otherwise
      max_ratio: the maximum aspect ratio that won't trigger the chunk
      target_ratio: when crops are chunked, they will be chunked to match this aspect ratio
      dilation: the width dilation of final chunks (to provide some overlaps)
      channels_last: whether the numpy array has dimensions in channels last order

  Returns:
      a tuple with the new crops, their mapping, and a boolean specifying whether any remap is required
  """
  _remap_required = False
  crop_map: list[int | tuple[int, int]] = []
  new_crops: list[np.ndarray] = []
  for crop in crops:
      # TODO: try to autocalculate target_ratio / max_ratio and dilation
      h, w = crop.shape[:2] if channels_last else crop.shape[-2:]
      aspect_ratio = w / h
      if aspect_ratio > max_ratio:
          # Determine the number of crops, reference aspect ratio = 4 = 128 / 32
          num_subcrops = int(aspect_ratio // target_ratio)
          # Find the new widths, additional dilation factor to overlap crops
          width = dilation * w / num_subcrops
          centers = [(w / num_subcrops) * (1 / 2 + idx) for idx in range(num_subcrops)]
          # Get the crops
          if channels_last:
              _crops = [
                  crop[:, max(0, int(round(center - width / 2))) : min(w - 1, int(round(center + width / 2))), :]
                  for center in centers
              ]
          else:
              _crops = [
                  crop[:, :, max(0, int(round(center - width / 2))) : min(w - 1, int(round(center + width / 2)))]
                  for center in centers
              ]
          # Avoid sending zero-sized crops
          _crops = [crop for crop in _crops if all(s > 0 for s in crop.shape)]
          # Record the slice of crops
          crop_map.append((len(new_crops), len(new_crops) + len(_crops)))
          new_crops.extend(_crops)
          # At least one crop will require merging
          _remap_required = True
      else:
          crop_map.append(len(new_crops))
          new_crops.append(crop)

  return new_crops, crop_map, _remap_required
```

f you find a solution that works well, feel free to open a PR so we can take a look. 👍

**Comment by Razlaw** [comment]
Wow, thanks for the detailed answer :)
I'll investigate a little more then 👍 

Some insights so far:
I used the minimal example you provided for inference on some other crops.
Original crop:

![Image](https://github.com/user-attachments/assets/94871c61-810c-4f2d-ab0d-76ec982dd117)

Result:
'2025-02-20T06:**51:1:13**.000Z'

The crop is split, but there might be something off with the merging as well.
I get similar results for another date time crop and the result you provided for the random string has a also a duplication in it.
XUrA+b3iHQP+F3qxu**0zzlZZ**DRA/6Z8+6nbf

Here are the recognized characters for the date time from above:
('2025-02-20', '2-20T06:51:1', '1:13.000Z')

And the subcrops right before recognition:

![Image](https://github.com/user-attachments/assets/42a2f146-e2e6-4be8-9f61-ae6e41d5a619)

![Image](https://github.com/user-attachments/assets/79240ac0-f2de-4645-8657-5c7477b947f7)

![Image](https://github.com/user-attachments/assets/ea353cac-89cf-4573-8803-a48601060a7c)

Long story short, I'll just look into the merging as well :)

**Comment by felixdittrich92** [comment]
Sounds good 👍🏼
And as mentioned if you should find a more robust way feel free to open a PR that we can have a look together 🙂
